llm/server.go

```diff
@@ -217,15 +217,35 @@ func NewLlamaServer(gpus discover.GpuInfoList, model string, ggml *GGML, adapter
 		params = append(params, "--threads", strconv.Itoa(defaultThreads))
 	}
 
-	flashAttnEnabled := envconfig.FlashAttention()
+	// Check flash attention support
+	var isEmbeddingModel bool
+	if _, ok := ggml.KV()[fmt.Sprintf("%s.pooling_type", ggml.KV().Architecture())]; ok {
+		isEmbeddingModel = true
+	}
 
-	for _, g := range gpus {
-		// only cuda (compute capability 7+) and metal support flash attention
-		if g.Library != "metal" && (g.Library != "cuda" || g.DriverMajor < 7) {
-			flashAttnEnabled = false
+	// Check flash attention support
+	flashAttnRequested := envconfig.FlashAttention()
+	if flashAttnRequested && gpus.SupportsFlashAttention() && supportsFlashAttention(ggml) {
```

Comment:
This should be able to get the kvCacheType once from the function below in `kv_cache.go`

Reply:
Yeah nice idea, done!